### PR TITLE
Feature/1364/284 clock

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

Use a mockable `Clock` to get the time.

Removes calls to `Instant.now()`, `new Date()`, `System.currentTimeMillis()` from production code.

## Why it does that

- Make access to system time consistent in code base.
- Make time mockable for testing.
- Make tests simple and reliable (avoid approximate time comparisons, etc.).

## Further notes

- Moved convenience method implementations from `DefaultServiceExtensionContext` to the interface
- Changed `DataPlaneInstance` contract to return null instead of 0 for empty `lastActive` time.
- `nbf`: in discussion with Ben
- Removed `rollbackState` method in ContractNegotiation and TransferProcess, as it's unused and violates state machine contract.
- Consistently inject services in `ContractServiceExtension`

## Linked Issue(s)

#284

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/jwt/DefaultJwtDecorator.java [hunk 1]
```diff
@@ -46,8 +48,7 @@ public void decorate(JWSHeader.Builder header, JWTClaimsSet.Builder claimsSet) {
                 .issuer(clientId)
                 .subject(clientId)
                 .jwtID(UUID.randomUUID().toString())
-                .notBeforeTime(new Date())
-                .issueTime(new Date())
-                .expirationTime(Date.from(Instant.now().plusSeconds(expiration)));
+                .issueTime(Date.from(clock.instant()))
```
Comment (chrislomonico) [inline extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/jwt/DefaultJwtDecorator.java:51]: nit: consistency - why not use 
`private final Instant createTime = clock.instant();` 
?

File: extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLease.java [hunk 2]
```diff
@@ -36,7 +37,7 @@ public void setLeaseId(String leaseId) {
         this.leaseId = leaseId;
     }
 
-    public boolean isExpired() {
-        return getLeasedAt() + getLeaseDuration() < Instant.now().toEpochMilli();
+    public boolean isExpired(Clock clock) {
+        return getLeasedAt() + getLeaseDuration() < clock.millis();
```
Comment (chrislomonico) [inline extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLease.java:41]: should this be `clock.toEpochMilli()`?

File: extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLeaseContext.java [hunk 3]
```diff
@@ -71,11 +73,11 @@ public void breakLease(String entityId) {
     @Override
     public void acquireLease(String entityId) {
         trxContext.execute(() -> {
-            var now = Instant.now().toEpochMilli();
+            var now = clock.millis();
```
Comment (chrislomonico) [inline extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLeaseContext.java:76]: Should this be `clock.toEpochMilli()`?

File: extensions/sql/transfer-process-store-sql/src/main/java/org/eclipse/dataspaceconnector/sql/transferprocess/store/SqlTransferProcessStore.java [hunk 4]
```diff
@@ -53,22 +53,23 @@ public class SqlTransferProcessStore implements TransferProcessStore {
...
-        leaseContext = SqlLeaseContextBuilder.with(transactionContext, leaseHolderName, statements);
+        this.clock = clock;
+        leaseContext = SqlLeaseContextBuilder.with(transactionContext, leaseHolderName, statements, clock);
     }
 
     @Override
     public @NotNull List<TransferProcess> nextForState(int state, int max) {
         var list = new ArrayList<TransferProcess>();
-        var now = Instant.now().toEpochMilli();
+        var now = clock.millis();
```
Comment (chrislomonico) [inline extensions/sql/transfer-process-store-sql/src/main/java/org/eclipse/dataspaceconnector/sql/transferprocess/store/SqlTransferProcessStore.java:72]: should this use `clock.toEpochMilli()`?

**Review by chrislomonico — CHANGES_REQUESTED** [review]
some general comments/questions.

File: extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/jwt/DefaultJwtDecorator.java (hunk 1 above)
Comment (algattik) [reply extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/jwt/DefaultJwtDecorator.java:51]: We want to get the current time every time (used to define token expiration) the method is called, not once when the application starts up

File: extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLease.java (hunk 2 above)
Comment (algattik) [reply extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLease.java:41]: That method doesn't exist. https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html#millis--

Comment (chrislomonico) [reply extensions/sql/lease-sql/src/main/java/org/eclipse/dataspaceconnector/sql/lease/SqlLease.java:41]: based on coversation, clock.Instant.toMilli is the same as clock.millis

**Review by chrislomonico — APPROVED** [review]
clarification provided, approving

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java [hunk 5]
```diff
@@ -134,14 +139,15 @@ private void registerServices(ServiceExtensionContext context) {
...
-        var validationService = new ContractValidationServiceImpl(agentService, definitionService, assetIndex, policyStore);
+        var validationService = new ContractValidationServiceImpl(agentService, definitionService, assetIndex, policyStore, clock);
         context.registerService(ContractValidationService.class, validationService);
 
         var waitStrategy = context.hasService(NegotiationWaitStrategy.class) ? context.getService(NegotiationWaitStrategy.class) : new ExponentialWaitStrategy(DEFAULT_ITERATION_WAIT);
 
         CommandQueue<ContractNegotiationCommand> commandQueue = new BoundedCommandQueue<>(10);
         CommandRunner<ContractNegotiationCommand> commandRunner = new CommandRunner<>(commandHandlerRegistry, monitor);
 
+        var clock = context.getClock();
```
Comment (cpeeyush) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java:150]: can we directly use injected clock here ?

File: extensions/data-plane-selector/selector-spi/src/test/java/org/eclipse/dataspaceconnector/dataplane/selector/strategy/RandomSelectionStrategyTest.java [hunk 6]
```diff
@@ -15,60 +15,23 @@
...
-            public long getLastActive() {
-                return 0;
-            }
-
-            @Override
-            public Map<String, Object> getProperties() {
-                return null;
-            }
-        };
+        instances = Collections.nCopies(1000, mock(DataPlaneInstance.class));
```
Comment (cpeeyush) [inline extensions/data-plane-selector/selector-spi/src/test/java/org/eclipse/dataspaceconnector/dataplane/selector/strategy/RandomSelectionStrategyTest.java:34]: is this change related to mockable clock work in this PR ? just pointing in case if this was accidently added here.

Comment (algattik) [reply extensions/data-plane-selector/selector-spi/src/test/java/org/eclipse/dataspaceconnector/dataplane/selector/strategy/RandomSelectionStrategyTest.java]: thanks, was related to another change which I reverted on, so I've rolled back changes on this class to for consistency

File: extensions/iam/decentralized-identity/identity-did-service/src/main/java/org/eclipse/dataspaceconnector/identity/DecentralizedIdentityServiceExtension.java [hunk 7]
```diff
@@ -54,7 +58,7 @@ public String name() {
     @Override
     public void initialize(ServiceExtensionContext context) {
         var vcProvider = createSupplier(context);
-        var identityService = new DecentralizedIdentityService(vcProvider, resolverRegistry, credentialsVerifier, context.getMonitor());
+        var identityService = new DecentralizedIdentityService(vcProvider, resolverRegistry, credentialsVerifier, context.getMonitor(), context.getClock());
```
[the remark this answers is not in the document]
Comment (algattik) [reply extensions/iam/decentralized-identity/identity-did-service/src/main/java/org/eclipse/dataspaceconnector/identity/DecentralizedIdentityServiceExtension.java:61]: Great point, fixed and also made consistent in other services (like Telemetry)

File: extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java [hunk 8]
```diff
@@ -38,12 +38,18 @@
 
 import static java.lang.String.format;
 
-// has to be "public", otherwise quartz won't be able to access is
 public class CrawlerJob implements Job {
 
     private static final String DIDS_PATH = "dids";
+    private static final Clock CLOCK = Clock.systemUTC();
```
Comment (cpeeyush) [inline extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java:44]: we can add method to `CrawlerContext` builder to add Clock. And then while creating Crawler config(org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerExtension.java:93) we can add the clock from `ServiceExtensionContext` 

```
var crawlerConfig = CrawlerContext.Builder.create()
               .....
                .clock(context.getClock())
                .build();
```

File: extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java [hunk 9]
```diff
@@ -54,7 +60,7 @@ public void execute(JobExecutionContext context) {
 
         monitor.info("CrawlerJob: browsing ION to obtain GaiaX DIDs");
 
-        var start = Instant.now();
+        var start = CLOCK.instant();
```
Comment (cpeeyush) [inline extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java:63]: based on the above comment then we will be able to get clock as `cc.clock()`.

File: extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java [hunk 10]
```diff
@@ -109,7 +115,8 @@ private List<String> getDidSuffixesForType(String type, TypeManager typeManager)
         try (var response = client.newCall(request).execute()) {
             if (response.isSuccessful()) {
                 var json = Objects.requireNonNull(response.body()).string();
-                var typeReference = new TypeReference<List<String>>() {};
+                var typeReference = new TypeReference<List<String>>() {
+                };
```
Comment (cpeeyush) [inline extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java:119]: nit: maybe not needed.
```suggestion
                var typeReference = new TypeReference<List<String>>() {};
```

File: extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java (hunk 8 above)
Comment (algattik) [reply extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java:44]: This class is instantiated by Quartz through the no-args constructor

Comment (cpeeyush) [reply extensions/iam/decentralized-identity/registration-service/src/main/java/org/eclipse/dataspaceconnector/iam/registrationservice/crawler/CrawlerJob.java:44]: Not sure if I understood it correctly because there are other values which crawler job gets from crawler context so in the same way we can have the clock as well.

See below comment for some more details

**Review by cpeeyush — APPROVED** [review]
